# Trabajo Práctico Integrador



## Urban Flow

## Integrantes del grupo:

* Salcedo Marcelo
* Gaspar Giannitrapani
* Ivan Porcari
* Rodrigo Zamora


## Sprint 1

## Consideraciones Generales e Importantes

- Desarrollo base: Los puntos solicitados son la base del trabajo práctico. Es posible la ampliación sin restricciones del mismo.

- Organización del Notebook:
  - Agregar antes del ejercicio 01 una celda donde se importan todas librerías necesarias.

  - Utilizar un título para cada celda y escribir cada funcionalidad en una celda con título. Tener en cuenta agrupar las funcionalidades afines.

## Versionado y Trabajo en Equipo

- La entrega puede entregarse en grupos de máximo 6 personas (mínimo 1 / máximo 6) y la defensa es individual. Los grupos deben mantenerse a lo largo del trabajo integrador.

- Se debe considerar que es un trabajo en equipo y que cualquiera de los integrantes puede realizar modificaciones para cada día de trabajo.

- Control de Versiones: Utilizar herramientas de versionado. Vincularlo por token a tu cuenta personal de versionado en la nube.

- Colaboradores: Brindar acceso al repositorio a las siguientes cuentas:

  - lcd-sa182@ugr.edu.ar

  - fpasinato@ugr.edu.ar

- Changelog: Cada ejercicio/punto del trabajo práctico es considerado como un día distinto de trabajo. Debe quedar constancia en un archivo CHANGELOG.md de los cambios generados cada día.

- Agregar una variable llamada desactivar_git_push que solamente me permita desactivar los "git push".  

- El archivo README.md debe contener el sprint en el que nos encontramos y debe contener:
  - El objetivo.
  - La introducción y el contexto del sprint de trabajo.


## Criterios de evaluación

- Funcionalidad:

  - El código debe ejecutar sin errores al ejecutar "Ejecutar todo" (Ctrl+F9).

  - Se debe implementar la totalidad de las funcionalidades solicitadas.
  
  - La lógica del programa debe ser correcta y consistente.
  
  - Utilización de librerías acorde a los problemas a resolver.

  - Recordar antes de entregar de validar el notebook con los siguientes pasos. Ir al menú Runtine/Ejecutar -> Restart session and run all/Reiniciar la sessión y ejecutar todo

- Claridad y organización:

  - El código debe estar bien organizado y dividido.

  - Utilizar diferentes celdas de código para resolver cada solicitud/punto dentro de un ejercicio específico con su comentario "title" correspondiente a la solución implementada.
  
  - Se deben utilizar nombres de variables que describan su contenido/función.
  
  - Se debe agregar comentarios explicativos y necesarios según estandares.

  - Todos los imports deben ubicarse en la primer celda de ejecución arriba de la descripción del "Ejercicio 01".

  - No se admite la sentencia "git add ." para simplificar el traceo.
  
  - No se admiten comentarios innecesarios dentro del código. Solo comentar funciones, clases, métodos y propiedades de clase.

- Estilo y presentación:

  - Se debe usar una indentación consistente y un formato de código adecuado.

  - Se deben evitar comentarios redundantes o irrelevantes.

  - Se debe presentar el código de forma ordenada y profesional.
  
  - Se debe utilizar [type hints](https://docs.python.org/3/library/typing.html).
  
  - Se deben utilizar comentarios según el [pep8](https://peps.python.org/pep-0008/).

- Herramientas:
  
  - Utilizar principalmente librerías vistas durante la cursada. No se admiten librerías similares a las vistas en la materia.

  - Utilizar librerías extras con criterios fundamentados, es decir, no utilizar por utilizar.

  - Indentar a 2 espacios y cortar a 80 caracteres según estandares.

  - No se puede utilizar gdrive.
  
  - No se pueden utilizar herramientas alternativas a las vistas durante la cursada.

## Condiciones de Entrega
- Formato de Archivo: Los estudiantes deberán enviar un archivo con extensión .ipynb.

- Nomenclatura: El nombre original del archivo debe reemplazar los valores indicados. Ejemplo: 0X_Price_Manager-Apellidos_Nombres.ipynb.

- Estado de la entrega:
  - Deberá entregarse ejecutado y con los datos precargados.
  - Deberá entregarse el desarrollo sobre una copia de la plantilla original.

- Canal de Entrega: La entrega se realiza en el aula virtual y solo se aceptan archivos (cualquier otra forma se considerará inválida). Solo se recibirán trabajos por correo adjunto a lcd-sa182@ugr.edu.ar si el plazo en la plataforma ha finalizado. No se aceptarán enlaces sueltos.


## Recomendaciones y Consecuencias del Incumplimiento
- Recomendaciones: Comenzar a trabajar con suficiente antelación. Leer y comprender el enunciado en su totalidad. Realizar consultas ante dudas al correo: lcd-sa182@ugr.edu.ar o fpasinato@ugr.edu.ar. La entrega en tiempo y forma es fundamental.

- Incumplimiento: Los trabajos que no cumplan con los requisitos no serán tomados en cuenta como nota final. Solo serán corregidos y evaluados con nota para la devolución correspondiente. Se considera incumplimiento a:

  - Entregas fuera de término.

  - Trabajos que no se entreguen sobre la plantilla de trabajo estipulada.

  - Trabajos que no se ejecuten completos (que contengan errores intermedios al dar "Ejecutar todo").

---

## Objetivo

El objetivo principal de este proyecto es aplicar los conocimientos adquiridos en para el versionado de código, la organización, limpieza del código y la utilización de pandas.


## Introducción y Contexto del problema


### Sprint 1

La localidad llamada Vaalserberg de Bélgica se encuentra en la zona fronteriza y limita con los paises de Países Bajos y Alemania. Esta localidad cuenta con un sistema de radares urbanos para la detección de infracciones por exceso de velocidad. Los registros históricos provienen de sistemas heredados, el cuál presenta errores de formato, faltante de datos generando registros inconsistentes en el nuevo sistema.

Debemos analizar y depurar los datos de viejo sistema para obtener información relevante sobre las infracciones y de está forma en el futuro poder incorporar los datos al nuevo sistema sin inconsistencias.

Descargar el dataset [speeding fines](https://raw.githubusercontent.com/HAD141/datasets/refs/heads/main/TrabajosPracticos/urban_flow/speeding_fines.csv) https://raw.githubusercontent.com/HAD141/datasets/refs/heads/main/TrabajosPracticos/urban_flow/speeding_fines.csv

El dataset contiene información histórica de multas por exceso de velocidad y presenta errores que deberán ser tratados para evitar inconsistencias.


---

## Ejercicio 01

> Puntos: 1

Inicialización y configuración de la herramienta de versionado. Vamos a trabajar sobre la rama llamada "Sprint_1".

La estructura de directorios es:

```
├── urban_flow
│   ├── data
│   │   ├── interim (almacena dataset procesados de pasos intermedios)
│   │   │   ├── plots
│   │   ├── processed (almacena dataset procesados finales para entregar a otra aplicación)
│   │   ├── raw (almacena los dataset en crudo, sin procesar)
```


In [1]:
# Imports para el proyecto
import pandas as pd
import numpy as np
import subprocess
import re
import matplotlib.pyplot as plt
import os
from pathlib import Path

In [2]:
desactivar_git_push: bool = False
# En Colab: activar token y reemplazar el remote URL
# from google.colab import userdata
# GITHUB_TOKEN = userdata.get('GITHUB_TOKEN')
# REMOTE_URL = f"https://{GITHUB_TOKEN}@github.com/Ivan-Porcari/tpi_hds_urban_flow.git"

REMOTE_URL: str = (
  "https://github.com/Ivan-Porcari/tpi_hds_urban_flow.git"
)

CARPETAS: list[str] = [
  "urban_flow/data/interim/plots",
  "urban_flow/data/processed",
  "urban_flow/data/raw",
]

In [3]:
# Git: init
if not os.path.exists(".git"):
  os.system("git init")
  os.system(
    "git config --local user.email 'ivanporcari.dev@gmail.com'"
  )
  os.system(
    "git config --local user.name 'Ivan Porcari'"
  )

In [4]:
# Git: rama Sprint_1 
ramas = subprocess.getoutput("git branch").replace("*", "").split()
if "Sprint_1" not in ramas:
  os.system("git checkout -b Sprint_1")
else:
  os.system("git checkout Sprint_1")

In [5]:
# Git: remote
remotes = subprocess.getoutput("git remote")
if "origin" not in remotes:
  os.system(f"git remote add origin {REMOTE_URL}")

In [6]:
# Estructura de carpetas 
for carpeta in CARPETAS:
  if not os.path.exists(carpeta):
    os.makedirs(carpeta)

In [7]:
# ── .gitignore ───────────────────────────────────────────────────────────────
GITIGNORE: str = """# Jupyter
.ipynb_checkpoints/

# Python
__pycache__/
*.pyc
*.pyo

# VS Code
.vscode/
"""

if not os.path.exists(".gitignore"):
  with open(".gitignore", "w", encoding="utf-8") as f:
    f.write(GITIGNORE)

In [8]:
# README.md 
README: str = """# Urban Flow

## Sprint actual: Sprint 1

## Objetivo
Aplicar conocimientos de versionado de código, organización, limpieza de datos y uso de pandas sobre un dataset de multas por exceso de velocidad.

## Introducción y contexto
La localidad de Vaalserberg (Bélgica), ubicada en zona fronteriza con Países Bajos y Alemania, cuenta con radares urbanos para detectar infracciones de velocidad. Los registros históricos provienen de sistemas heredados con errores de formato y datos faltantes. 
El objetivo del Sprint 1 es analizar y depurar esos datos para incorporarlos al nuevo sistema sin inconsistencias.
"""

if not os.path.exists("README.md"):
  with open("README.md", "w", encoding="utf-8") as f:
    f.write(README)

In [9]:
# CHANGELOG.md 
CHANGELOG: str = """# Changelog

Todos los cambios notables del proyecto Urban Flow 
serán documentados en este archivo.
El formato está basado en 
[Keep a Changelog](https://keepachangelog.com).

## [0.1.0] - 2026-04-16

### Added
- Inicialización del repositorio git
- Configuración del archivo .gitignore
- Creación del archivo README.md
- Creación de la estructura de carpetas del proyecto
"""

if not os.path.exists("CHANGELOG.md"):
  with open("CHANGELOG.md", "w", encoding="utf-8") as f:
    f.write(CHANGELOG)

In [10]:
# Commit ejercicio 01
os.system("git add README.md")
os.system("git add CHANGELOG.md")
os.system("git add .gitignore")
os.system("git add urban_flow/data/raw/.gitkeep")
os.system("git add urban_flow/data/processed/.gitkeep")
os.system("git add urban_flow/data/interim/.gitkeep")
os.system("git add urban_flow/data/interim/plots/.gitkeep")
os.system(
  'git commit -m "Ejercicio 01: init repo, estructura y archivos base"'
)

if not desactivar_git_push:
  os.system("git push -u origin Sprint_1")

In [11]:
# Archivos .gitkeep para trackear carpetas vacías
for carpeta in CARPETAS:
  gitkeep = os.path.join(carpeta, ".gitkeep")
  if not os.path.exists(gitkeep):
    open(gitkeep, "w").close()

In [12]:
os.system("git add urban_flow/data/raw/.gitkeep")
os.system("git add urban_flow/data/processed/.gitkeep")
os.system("git add urban_flow/data/interim/.gitkeep")
os.system("git add urban_flow/data/interim/plots/.gitkeep")
os.system(
  'git commit -m "Ejercicio 01: agrego .gitkeep en carpetas vacías"'
)

if not desactivar_git_push:
  os.system("git push origin Sprint_1")

In [13]:
print(subprocess.getoutput("git status"))

On branch Sprint_1
Your branch is up to date with 'origin/Sprint_1'.

nothing to commit, working tree clean


In [14]:
print(subprocess.getoutput("git log --oneline"))

92bc432 Agregamos funcionalidad Merge
4981f0c  Fix : Ejercicio3 : Normalizacion y limpieza de dataset/ - Generacion de nuevo dataset limpio en interim
f22859c Agregamos funcionalidad Merge
6de06e6 Agregamos funcionalidad Merge
5fd0d0b Ejercicio3 : Normalizacion y limpieza de dataset - Generacion de nuevo dataset limpio en interim
d88991d Ejercicio2 : Incorporación de dataset original, Implementación de carga de dataset, analisis de tipos, valores faltantes y conteo de nulos.
904690b Ejercicio 01: agrego notebook al repositorio
fd1176b Ejercicio 01: agrego .gitkeep en carpetas vacías
c4cb7a4 Ejercicio 01: init repo, estructura y archivos base


In [15]:
# Push manual commit pendiente
if not desactivar_git_push:
  os.system("git push origin Sprint_1")

## Ejercicio 02

> Puntos: 1

Los desarrollos que deben realizar son:

- Descargar el dataset raw original y almacenarlo en `urban_flow/data/raw`.

- Mostrar las 5 primeras filas.

- Analizar tipos de datos

- Contar los valores nulos.



##### Creo la rama ejercicio_2_3

In [16]:
RAMA_ACTUAL: str = "ejercicio_2_3"

ramas = subprocess.getoutput("git branch").replace("*", "").split()


if RAMA_ACTUAL not in ramas:
  # Si la rama no existe, se crea  y se salta a ella automáticamente
  os.system(f"git checkout -b {RAMA_ACTUAL}")
else:
  # Si la rama ya existe, simplemente se cambia a ella
  os.system(f"git checkout {RAMA_ACTUAL}")

##### Configurar Ruta Relativa

In [17]:

BASE_DIR = Path.cwd() # Ruta Base

BASE_RAW = BASE_DIR / "urban_flow"/"data"/"raw"  # Ruta hacia los datos crudos

##### Creando dataframe con el archivo .csv

In [18]:
df = pd.read_csv(BASE_RAW / "speeding_fines.csv")

##### Mostrar las 5 primeras filas

In [19]:
print(df.head(5))

   multa_id    patente         fecha          hora  velocidad_registrada  \
0         1    FD 1475    24/09/2021      09:01 AM                   NaN   
1         2   0695 DVH    19/05/2020         18:01                   NaN   
2         3   WU 341 A    2022/07/17      09:01 AM                 107.0   
3         4   GN64 OTP  invalid_date  invalid_hour                   NaN   
4         5  BX 55 DOG    2021/01/21  invalid_hour                  94.0   

   velocidad_maxima         ubicacion radar_id estado_multa  
0                60  Av. Siempre Viva      R01      APELADA  
1                40   Av Siempre Viva      NaN      APELADA  
2                60     AV LIBERTADOR      R03       PAGADA  
3                60            Ruta 9      NaN      APELADA  
4                60  Av. Siempre Viva      NaN        ERROR  


##### Analizar Tipos de Datos

In [20]:
print(df.dtypes)

multa_id                  int64
patente                  object
fecha                    object
hora                     object
velocidad_registrada    float64
velocidad_maxima          int64
ubicacion                object
radar_id                 object
estado_multa             object
dtype: object


##### Contar los valores nulos

In [21]:
#Identificar las columnas con datos faltantes 

print(df.isnull().sum())

multa_id                   0
patente                   50
fecha                      0
hora                       0
velocidad_registrada    1957
velocidad_maxima           0
ubicacion                  0
radar_id                1002
estado_multa               0
dtype: int64


##### Modificacion de CHANGELOG

In [22]:
CAMBIOS: str = """
## [0.2.0] - 2026-04-23

### Added
- Incorporación delarchivo 'speeding_fines.csv' a la carpeta raw del proyecto
- Carga de dataset 'speeding_fines.csv' mediante pathlib.
- Analisis de tipo (dtypes) y  valores nulos (isnull().sum()).
- Creación de la rama ejercicio_2_3
"""

#  Verificamos que el archivo existe y añadimos la información
if os.path.exists("CHANGELOG.md"):
  with open("CHANGELOG.md", "a", encoding="utf-8") as f:
    f.write(CAMBIOS)
  print("Changelog actualizado con éxito.")


Changelog actualizado con éxito.


##### Commit Ejercicio 2

In [23]:


os.system("git add CHANGELOG.md")
os.system("git add urban_flow/data/raw/speeding_fines.csv")
os.system("git add 01_Urban_Flow_Porcari_Ivan.ipynb")
os.system("git commit -m 'Ejercicio2 : Incorporación de dataset original, Implementación de carga de dataset, analisis de tipos, valores faltantes y conteo de nulos.'")


1

In [24]:


os.system("git add CHANGELOG.md")
os.system("git add urban_flow/data/raw/speeding_fines.csv")
os.system("git add 01_Urban_Flow_Porcari_Ivan.ipynb")
os.system("git commit -m 'Ejercicio2 : Incorporación de dataset original, Implementación de carga de dataset, analisis de tipos, valores faltantes y conteo de nulos.'")


1

##### Push Ejercicio 2

In [25]:
if not desactivar_git_push:
  os.system("git push -u origin ejercicio_2_3")

##### Informacion de la Rama

In [26]:
print(subprocess.getoutput("git status"))

On branch ejercicio_2_3
Your branch is up to date with 'origin/ejercicio_2_3'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   CHANGELOG.md



## Ejercicio 03

> Puntos: 2

Utilizando Pandas:

- Normalizar las de fechas con el formato 'YYYY-MM-DD', en caso de fechas inválidas completar con la fecha `1932-01-01`. Mostrar las 10 primeras fechas.

- Normalizar los horas al formato de 24hs, las horas inválidas se procesaran como `00:00`. Mostrar las 10 primeras horas.

- Normalizar las ubicaciones. Quitar caracteres especiales y pasar todo a mayúsculas. Mostrar las 10 primeras ubicaciones.

- Limpiar y normalizar las patentes, en caso de error completar con pd.NA (no disponible/not available). Quitar caracteres especiales y pasar todo a mayúsculas. Mostrar las 10 últimas patentes.

- Eliminar las filas que posean valores relevantes para las multas y que se encuentren vacios. Mostrar la cantidad de registros eliminados y cuales son las columnas que se observaron utilizando la leyenda
```
Se eliminaron X filas.
Las columnas observadas son:
 - XXX
 - XXX
 - ...
```

- Detectar y eliminar los outliers. Mostrar la cantidad de registros eliminados y cuales son las columnas que se observaron utilizando la leyenda
```
Se eliminaron X filas.
Las columnas observadas son:
 - XXX
 - XXX
 - ...
```

- Crear la columna `exceso_velocidad_real` y calcularla. El exceso de velocidad real es la diferencia entre velocidad_registrada y la velocidad_maxima. Mostrar las primeras 10 patentes y su exceso de velocidad real.

- Crear la columna `exceso_velocidad` y calcularla. El exceso de velocidad es la diferencia entre velocidad_registrada y la velocidad_maxima más un 5 porciento de la velocidad máxima. Mostrar las primeras 10 patentes y su exceso de velocidad.

- Eliminar las filas que no cometieron infracciones según el exceso de velocidad.Mostrar cuantas filas se eliminaron con la leyenda
```
Se eliminaron X filas.
```

- Guardar el dataset limpio en `urban_flow/data/interim/speeding_fines.csv`

##### Normalizar Fecha 

In [27]:
# Convierte la columna a fecha, si los datos son invalidos, los 
# transforma en NaT para no generar un error
#
  
df['fecha'] = pd.to_datetime(df['fecha'], errors='coerce', format='mixed')

df['fecha'] = df['fecha'].fillna('1932-01-01') # Podria usar pd.Timestamp('1900-01-01)

print(df['fecha'].head(10))



0   2021-09-24
1   2020-05-19
2   2022-07-17
3   1932-01-01
4   2021-01-21
5   2023-07-27
6   2023-06-16
7   2023-01-11
8   1932-01-01
9   1932-01-01
Name: fecha, dtype: datetime64[ns]


##### Normalizar Hora

In [28]:
# Convierte la columna a hora, busca horas en diferentes formatos y 
# los incorrectos los convierte en nulos
df['hora'] = pd.to_datetime(df['hora'], errors='coerce', format='mixed')

# Aplica a las horas un formato de 24hs sin segundos
# Rellena los nulos con 00:00 
df['hora'] = df['hora'].dt.strftime('%H:%M').fillna('00:00')
print(df['hora'].head(10))

0    09:01
1    18:01
2    09:01
3    00:00
4    00:00
5    00:00
6    18:01
7    08:01
8    09:01
9    14:01
Name: hora, dtype: object


##### Funcion de limpieza de ubicaciones


In [29]:
def normalizar_ubicaciones(texto:str):

    if pd.isna(texto):
        return "Desconocida"
    
    # Buscar 'av' o 'av.' (ignora mayúsculas/minúsculas)
    # av -> busca letras a y v
    # \. -> quiero que el punto actue como punto
    texto = re.sub(r'(?i)av\.?\s*', 'AV ', texto)
    
    # Buscar cualquier cosa que no sea letra, número o espacio.
    texto = re.sub(r'[^a-zA-Z0-9\s]', '', texto)

    # Buscar espacios en blanco 
    texto= re.sub(r'\s+', ' ', texto).strip()

    #Retorno palabra en Mayusculas
    return texto.upper()

##### Mostrar las primera 10 ubicaciones

In [30]:
# Aplicamos la función a la columna
df['ubicacion'] = df['ubicacion'].apply(normalizar_ubicaciones)

# Mostrar las primeras 10  ubicaciones
print(df['ubicacion'].head(10))

0    AV SIEMPRE VIVA
1    AV SIEMPRE VIVA
2      AV LIBERTADOR
3             RUTA 9
4    AV SIEMPRE VIVA
5    AV SIEMPRE VIVA
6      AV LIBERTADOR
7    AV SIEMPRE VIVA
8    AV SIEMPRE VIVA
9      AV LIBERTADOR
Name: ubicacion, dtype: object


##### Limpieza y normalización de patentes


In [31]:
def normalizar_patentes(valor):

    # Si el valor es nulo de entrada, devolvemos pd.NA 
    if pd.isna(valor) or valor == "":
        return pd.NA
    
     # Quitar caracteres especiales
    texto_limpio = re.sub(r'[^a-zA-Z0-9]', '', valor)

    # Verificar si después de la limpieza quedó vacío (caso de error)
    if texto_limpio == "":
        return pd.NA

    # Retornar en mayusculas
    return texto_limpio.upper()

In [32]:
#Ejecutar la limpieza y normalizacion como string
df['patente'] = df['patente'].apply(normalizar_patentes).astype("string")


##### Mostrar las primera 10 patentes

In [33]:
print(df['patente'].head(10))

0     FD1475
1    0695DVH
2     WU341A
3    GN64OTP
4    BX55DOG
5       <NA>
6    CD05601
7      BP717
8     41RSTZ
9    OUTSTDN
Name: patente, dtype: string


##### Eliminacion de filas con valores relevantes para las multas

In [34]:
# Columnas con datos relevantes para generar multas
columnas_relevantes = [
    'patente', 
    'fecha', 
    'hora', 
    'velocidad_registrada', 
    'velocidad_maxima', 
    'ubicacion'
]
filas_totales_df= len(df)

# Se comentaron las siguientes dos lineas ya que si bien son datos relevantes 
# a la hora de realizar una multa, pero posteriormente se pide un gráfico soicitando
# estos datos
#df['fecha'] = df['fecha'].replace('1932-01-01', np.nan)
#df['hora'] = df['hora'].replace('00:00', np.nan)


# Solo buscamos en las columnas relevantes (subset), no en todo el df
# Eliminamos las filas con valores vacios
df = df.dropna(subset=columnas_relevantes)

# Restamos las filas actuales a las filas iniciales para saber 
# cuantas filas fueron eliminadas
filas_eliminadas = filas_totales_df - len(df)


print(f"Se eliminaron {filas_eliminadas} filas. Las columnas observadas son:")
for columna in columnas_relevantes:
    print(f"- {columna}")

Se eliminaron 1982 filas. Las columnas observadas son:
- patente
- fecha
- hora
- velocidad_registrada
- velocidad_maxima
- ubicacion


##### Detectar y eliminar outliers

In [35]:
# Definimos las columnas numéricas donde buscaremos los outliers
columnas_observadas = ['velocidad_registrada', 'velocidad_maxima']

# Guardo las filas antes de limpiar
filas_antes = len(df)

#  Aplicamos el método del rango intercuartilico (IQR) para cada columna
for columna in columnas_observadas:
    # Calculamos el primer cuartil (Q1) 
    # Calculamos el tercer cuartil (Q3) 
    # Calculamos el rango intercuartílico

    Q1 = df[columna].quantile(0.25)    
    Q3 = df[columna].quantile(0.75)    
    IQR = Q3 - Q1
    
    # Definimos los limites permitidos
    limite_inferior = Q1 - 1.5 * IQR
    limite_superior = Q3 + 1.5 * IQR
    
    # Conservamos solo las filas que esten dentro de los limites
    df = df[(df[columna] >= limite_inferior) & (df[columna] <= limite_superior)]

# Calculamos matemáticamente cuántas filas se eliminaron
filas_eliminadas = filas_antes - len(df)


print(f"Se eliminaron {filas_eliminadas} filas. Las columnas observadas son:")
for col in columnas_observadas:
    print(f"- {col}")

Se eliminaron 0 filas. Las columnas observadas son:
- velocidad_registrada
- velocidad_maxima


##### Exceso de velocidad real

In [36]:
# Creo la columna con la resta de la velocidad registrada y la velocidad maxima
df['exceso_velocidad_real'] = df['velocidad_registrada'] - df['velocidad_maxima']

# Seleccionamos solo las columnas patente y exceso_velocidad_real y tomamos las primeras 10
print(df[['patente', 'exceso_velocidad_real']].head(10))


    patente  exceso_velocidad_real
2    WU341A                   47.0
4   BX55DOG                   34.0
7     BP717                   10.0
9   OUTSTDN                   31.0
11  TL03GOG                   -2.0
12   GE1466                   35.0
16  5777770                   53.0
20  BHRP770                   11.0
21   RFH373                   41.0
23   XIIUME                   12.0


##### Exceso de velocidad

In [37]:
# Agregamos el 5% de velocidad a la velocidad maxima
velocidad_tolerada = df['velocidad_maxima'] * 1.05

# Creamos la columna exceso de velocidad y hacemos la resta
df['exceso_velocidad'] = df['velocidad_registrada'] - velocidad_tolerada

# 3. Filtramos las columnas pedidas y mostramos las 10 primeras
print(df[['patente', 'exceso_velocidad']].head(10))

    patente  exceso_velocidad
2    WU341A              44.0
4   BX55DOG              31.0
7     BP717               7.0
9   OUTSTDN              29.0
11  TL03GOG              -4.0
12   GE1466              33.0
16  5777770              51.0
20  BHRP770               8.0
21   RFH373              37.0
23   XIIUME              10.0


##### Eliminando las filas sin infracciones

In [38]:
# Guardamos la cantidad total de filas que hay en la tabla actualmente
filas_antes = len(df)

# Conservamos las filas con exceso de velocidad mayor a 0
df = df[df['exceso_velocidad'] > 0]

# Calculamos las filas eliminadas
filas_eliminadas = filas_antes - len(df)

print(f"Se eliminaron {filas_eliminadas} filas.")

Se eliminaron 305 filas.


##### Guardamos el Dataset Limpio

In [39]:
# Configuramos las rutas
BASE_INTERIM = BASE_DIR / "urban_flow"/"data"/"interim" 
ruta_save = BASE_INTERIM / "speeding_fines.csv"

# Exportamos el df a csv sin generar la columna N° filas
df.to_csv(ruta_save, index=False)

##### Modificación de CHANGELOG EJ 3

In [40]:
CAMBIOS: str = """
## [0.3.1] - 2026-04-24

### Added
- Normalización de fechas con formato YYYY-MM-DD
- Normalización de horas a formato 24 HS 
- Normalización de Ubicaciones, se quitaron caracteres especiales y se pasaron a mayúsculas
- Normalización de patentes, se quitaron caracteres especiales y pasaron a mayúsculas.
- Se eliminaron las filas que tenían inconsistencia de datos que imposibilitaban las generación de la infracción
- Se buscaron outliers en las columnas numericas
- Se creo la columna exceso_velocidad_real y exceso_velocidad
- Se eliminaron las filas con patentes que no cometieron infracciones.
- Se generó un dataset limpio en la carpeta interim

"""

#  Verificamos que el archivo existe y añadimos la información
if os.path.exists("CHANGELOG.md"):
  with open("CHANGELOG.md", "a", encoding="utf-8") as f:
    f.write(CAMBIOS)
  print("Changelog actualizado con éxito.")

Changelog actualizado con éxito.


##### Actualización CHANGELOG

In [41]:
CAMBIOS: str = """
## [0.3.1] - 2026-04-25

### Fixed
- Se corrigio el codigo para que los datos invalidos representados con 1932-01-01 y 00:00, queden en el dataset 
para su posterior uso.

"""

#  Verificamos que el archivo existe y añadimos la información
if os.path.exists("CHANGELOG.md"):
  with open("CHANGELOG.md", "a", encoding="utf-8") as f:
    f.write(CAMBIOS)
  print("Changelog actualizado con éxito.")

Changelog actualizado con éxito.


##### Commit Ej3

In [42]:


os.system("git add CHANGELOG.md")
os.system("git add urban_flow/data/interim/speeding_fines.csv")
os.system("git add 01_Urban_Flow_Porcari_Ivan.ipynb")
os.system("git commit -m ' Fix : Ejercicio3 : Normalizacion y limpieza de dataset/ - Generacion de nuevo dataset limpio en interim'")


1

##### Push Ejercicio 3

In [43]:
if not desactivar_git_push:
  os.system("git push -u origin ejercicio_2_3")

#### Merge para Ejercicios 2 y 3

In [44]:
# Agregamos al Stage y hacemos Commit
os.system("git add 01_Urban_Flow_Porcari_Ivan.ipynb") 
os.system("git commit -m 'Agregamos funcionalidad Merge'")
# Procedemos al Merge
os.system("git checkout Sprint_1")
os.system("git merge ejercicio_2_3")
os.system("git push origin Sprint_1")

0

## Ejercicio 04

> Puntos: 2

Definir una clase FineAnalyzer que cumpla con las siguientes funcionalidades:

- Reciba el DataFrame limpio como argumento de inicialización.

- Encapsule correctamente los datos.

- Implemente métodos para:

  - Ranking de las patentes más multadas (top 5). Debe retornar un DataFrame ordenado de mayor a menor y el índice debe comenzar desde 1, con las columnas de patentes y cantidad..

  - Ranking de los horarios donde se producen más multas (top 5). Debe retornar un DataFrame ordenado de mayor a menor y el índice debe comenzar desde 1, con las columnas de hora y cantidad.


  - Exceso promedio de velocidad. Debe retorna un flotante.

  - Exceso real promedio de velocidad . Debe retorna un flotante.

  - Contabilizar la cantidad de multas por cada  ubicación. Debe retornar un DataFrame ordenado por ubicación mostrando la ubicación y la cantidad.

- Crear el objecto e invocar cada uno de los métodos en celdas separadas.
  

## Punto 05

> Puntos: 2


Responder a cada punto con un gráfico.

- Ranking de las 10 patentes más reincidentes ordenadas de mayor a menor. Exportar el gráfico con el nombre `urban_flow/data/interim/plots/fines.jpg`


- Mostrar el porcentaje de infracciones por hora en un gráfico de torta. Exportar el gráfico con el nombre `urban_flow/data/interim/plots/hours.jpg`

- Mostrar la cantidad de infracciones por mes en un gráfico de barras horizontal ordenado de mayor a menor. Exportar el gráfico con el nombre `urban_flow/data/interim/plots/months.jpg`

- Mostrar en un gráfico de líneas de los excesos de velocidad agrupados por la hora 00:00. Exportar el gráfico con el nombre `urban_flow/data/interim/plots/hour.jpg`

- Mostrar en un gráfico de líneas de los excesos de velocidad agrupados por la fecha 1932-01-01. Exportar el gráfico con el nombre `urban_flow/data/interim/plots/date.jpg`

## Punto 06

> Puntos: 1

- Mostrar el porcentaje de infracción que se produjeron en la fecha 1932-01-01.

```
El porcentaje de infracciones en la fecha 1932-01-01 es XX.XX%
```

- Mostrar el porcentaje de infracción que se produjeron en la hora 00:00.
```
El porcentaje de infracciones a la hora 00:00 es XX.XX%
```

In [53]:
# Visualizo los 10 primeros registros del dataset limpio

df.head(10)

,multa_id,patente,fecha,hora,velocidad_registrada,velocidad_maxima,ubicacion,radar_id,estado_multa,exceso_velocidad_real,exceso_velocidad
2,3,WU341A,2022-07-17,09:01,107.0,60,AV LIBERTADOR,R03,PAGADA,47.0,44.0
4,5,BX55DOG,2021-01-21,00:00,94.0,60,AV SIEMPRE VIVA,NaN,ERROR,34.0,31.0
7,8,BP717,2023-01-11,08:01,70.0,60,AV SIEMPRE VIVA,R01,PAGADA,10.0,7.0
9,10,OUTSTDN,1932-01-01,14:01,71.0,40,AV LIBERTADOR,R02,APELADA,31.0,29.0
12,13,GE1466,2020-08-12,18:01,75.0,40,AV LIBERTADOR,NaN,IMPAGA,35.0,33.0
16,17,5777770,2020-01-29,09:02,93.0,40,AV SIEMPRE VIVA,R02,APELADA,53.0,51.0
20,21,BHRP770,1932-01-01,08:01,71.0,60,RUTA 9,R03,PAGADA,11.0,8.0
21,22,RFH373,2020-10-15,09:02,121.0,80,AV LIBERTADOR,R01,IMPAGA,41.0,37.0
23,24,XIIUME,2021-06-12,08:01,52.0,40,AV SIEMPRE VIVA,R03,IMPAGA,12.0,10.0
25,26,HF58HXP,2022-11-20,14:01,62.0,40,RUTA 9,NaN,APELADA,22.0,20.0


##### Porcentaje de infracciones con fecha 1932-01-01

Se calcula el porcentaje de infracciones registradas con la fecha 
1932-01-01. Esta fecha fue asignada en el Ejercicio 03 a todos los 
registros que tenían una fecha inválida en el sistema heredado. Se 
divide la cantidad de filas con ese valor sobre el total de filas 
del dataset limpio.

In [50]:
# Calculo la cantidad total de registros del dataset df
# Esto representa el 100% de las multas registradas
total_filas = len(df)

# Calculo la cantidad de registros con fecha 1932-01-01
filas_fecha_invalida = len(df[df['fecha'] == '1932-01-01'])

# Regla de 3 simple:
# total_filas          -> 100%
# filas_fecha_invalida -> x
porcentaje_fecha = (filas_fecha_invalida * 100) / total_filas

print(f"El porcentaje de infracciones en la fecha 1932-01-01 es {porcentaje_fecha:.2f}%")


El porcentaje de infracciones en la fecha 1932-01-01 es 26.44%


##### Porcentaje de infracciones con hora 00:00

Se calcula el porcentaje de infracciones registradas con la hora 
00:00. Este valor fue asignado en el Ejercicio 03 a todos los 
registros que tenían una hora inválida en el sistema heredado. Se 
divide la cantidad de filas con ese valor sobre el total de filas 
del dataset limpio.

In [51]:
# Calculo la cantidad total de registros del dataset
# Esto representa el 100% de las multas registradas
total_filas = len(df)

# Calculo la cantidad de registros con hora 00:00
filas_hora_invalida = len(df[df['hora'] == '00:00'])

# Regla de 3 simple:
# total_filas        -> 100%
# filas_hora_invalida -> x
porcentaje_hora = (filas_hora_invalida * 100) / total_filas

print(f"El porcentaje de infracciones a la hora 00:00 es {porcentaje_hora:.2f}%")


El porcentaje de infracciones a la hora 00:00 es 19.79%


## Punto 07

> Puntos: 1



Redactar una conclusión acerca de los datos que contiene el dataset.


El dataset original contaba con 4000 registros de multas por exceso 
de velocidad provenientes del sistema heredado de la localidad de 
Vaalserberg. Luego del proceso de limpieza y normalización, quedaron 
1713 registros válidos, lo que significa que se descartó más de la 
mitad de los datos por errores de formato, valores faltantes o 
registros que directamente no representaban una infracción real.

Durante la normalización se detectaron problemas importantes en los 
datos originales. Las fechas y horas presentaban formatos 
inconsistentes e inválidos, por lo que se asignaron valores centinela 
para identificarlos: la fecha 1932-01-01 para fechas inválidas y 
00:00 para horas inválidas. El análisis posterior mostró que el 
26.44% de las multas tienen fecha 1932-01-01 y el 19.79% tienen hora 
00:00, lo que indica que una parte considerable de los registros no 
tenía información temporal confiable en el sistema original.

También se detectaron patentes con formatos incorrectos, ubicaciones 
con caracteres especiales y registros sin velocidad registrada, todos 
los cuales fueron tratados durante el proceso de limpieza. Además, 
se eliminaron los registros donde el exceso de velocidad no superaba 
el 5% de tolerancia sobre la velocidad máxima permitida, ya que esos 
casos no constituyen una infracción real según la normativa aplicada.

En conclusión, el sistema heredado presentaba serias deficiencias en 
la carga y validación de datos. Los registros recuperados y limpios 
representan una base sólida para incorporar al nuevo sistema, pero 
sería recomendable revisar los procesos de carga de datos para evitar 
que estas inconsistencias se repitan en el futuro.

In [45]:
#%%writefile -a data/Readme.md
